### PR TITLE
Display suspension to user

### PR DESCRIPTION
**Pull request by imorland**
Addresses https://github.com/flarum/core/issues/1447

- [x] Expand `SuspendUserModal` to include two new optional fields: reason (for moderator use) and message (for display to the suspended user
- [x] Display a modal to a suspended user, displaying the message entered by the moderator, and the length of the suspension
- [x] Move `Form-group` components to an `ItemList` for extensibility
- [x] `compat` exports
- [x] Expand notification blueprints to include email notifications of suspension/unsuspension, including the suspension message

image

image

image

### CODE REVIEW

**Comment by askvortsov1** [comment]
This is excellent, but will probably need to wait till the next release for review since we need to prioritize milestone items and fixed right now. Thanks for putting it together!

File: js/src/forum/components/SuspendUserModal.js [hunk 1]
```diff
@@ -95,10 +138,10 @@ export default class SuspendUserModal extends Modal {
         break;
 
       default:
-        // no default
+      // no default
     }
 
-    this.attrs.user.save({suspendedUntil}).then(
+    this.attrs.user.save({ suspendedUntil: suspendedUntil, suspendReason: this.reason(), suspendMessage: this.message() }).then(
```
Comment (davwheat) [inline js/src/forum/components/SuspendUserModal.js]:
```suggestion
    this.attrs.user.save({ suspendedUntil, suspendReason: this.reason(), suspendMessage: this.message() }).then(
```

File: js/src/forum/components/SuspendUserModal.js [hunk 2]
```diff
@@ -79,12 +55,79 @@ export default class SuspendUserModal extends Modal {
...
+        {this.status() === 'limited' ? (
+          <div className="SuspendUserModal-days-input">
+            <input type="number"
+              min="0"
+              value={this.daysRemaining()}
+              oninput={withAttr('value', this.daysRemaining)}
+              className="FormControl" />
+            {app.translator.trans('flarum-suspend.forum.suspend_user.limited_time_days_text')}
+          </div>
+        ) : ''}
```
Comment (davwheat) [inline js/src/forum/components/SuspendUserModal.js:98]: A little nicer than a ternary

```suggestion
        {this.status() === 'limited' && (
          <div className="SuspendUserModal-days-input">
            <input type="number"
              min="0"
              value={this.daysRemaining()}
              oninput={withAttr('value', this.daysRemaining)}
              className="FormControl" />
            {app.translator.trans('flarum-suspend.forum.suspend_user.limited_time_days_text')}
          </div>
        )}
```

File: locale/en.yml [hunk 3]
```diff
@@ -13,17 +13,25 @@ flarum-suspend:
 
   # Translations in this namespace are used by the forum user interface.
   forum:
+    # These translations are used in the suspension info modal
+    infomodal:
+      dismiss_button: Dismiss
+      indefinite: This is an indefinite suspension
+      limited: "This suspension will be in force until {date}"
+      title: This account is suspended
     # These translations are used in the suspension notifications
```
Comment (davwheat) [inline locale/en.yml:22]: Just matches the existing standard of blank lines between sections
```suggestion
      title: This account is suspended

    # These translations are used in the suspension notifications
```

File: js/src/forum/components/SuspendUserModal.js [hunk 4]
```diff
@@ -3,12 +3,15 @@ import Button from 'flarum/components/Button';
...
 import withAttr from 'flarum/utils/withAttr';
+import ItemList from 'flarum/utils/ItemList';
 
 export default class SuspendUserModal extends Modal {
   oninit(vnode) {
     super.oninit(vnode);
 
     let until = this.attrs.user.suspendedUntil();
+    let reason = this.attrs.user.suspendReason();
+    let message = this.attrs.user.suspendMessage();
```
Comment (davwheat) [inline js/src/forum/components/SuspendUserModal.js:16]: Can these be `const`s?

File: js/src/forum/components/SuspendUserModal.js (hunk 2 above)
Comment (davwheat) [inline js/src/forum/components/SuspendUserModal.js]: A little neater than a ternary :)
```suggestion
        {this.status() === 'limited' && (
          <div className="SuspendUserModal-days-input">
            <input type="number"
              min="0"
              value={this.daysRemaining()}
              oninput={withAttr('value', this.daysRemaining)}
              className="FormControl" />
            {app.translator.trans('flarum-suspend.forum.suspend_user.limited_time_days_text')}
          </div>
        )}
```

File: js/src/forum/components/SuspendUserModal.js [hunk 5]
```diff
@@ -3,12 +3,15 @@ import Button from 'flarum/components/Button';
...
     let status = null;
 
     if (new Date() > until) until = null;
```
Comment (askvortsov1) [inline js/src/forum/components/SuspendUserModal.js:19]: Could this be a ternary?

File: locale/en.yml [hunk 6]
```diff
@@ -35,3 +43,25 @@ flarum-suspend:
     # These translations are found on the user profile page (admin function).
     user_controls:
       suspend_button: Suspend
+
+  # Translations in this namespace are used by suspension email notifications
+  email:
+    suspended:
+      subject: You have been suspended
```
Comment (askvortsov1) [inline locale/en.yml:50]: "Your account has been suspended"?

File: locale/en.yml [hunk 7]
```diff
@@ -35,3 +43,25 @@ flarum-suspend:
...
+      body: |
+        Hey {recipient_display_name},
+
+        You've been suspended for the following reason:
```
Comment (askvortsov1) [inline locale/en.yml:54]: Should include a bit more context: "suspended from ______", maybe a duration?

**Review by askvortsov1 — CHANGES_REQUESTED** [review]

**Comment by dsevillamartin** [comment]
This PR now includes commits that aren't relevant to the issue it's trying to fix.

**Comment by davwheat** [comment]
It might be worth copying the gg branch to this repo, then recreating the PR so we could actually merge this.

**Comment by askvortsov1** [comment]
@imorland is this something you're still interested in working on? If not, we might want to close this for now

**Comment by imorland** [comment]
@askvortsov1 I'd absolutely like to get this completed. Unfortunately this was started during my time at giffgaff, so I'm not able to make any changes to this PR now.

As @davwheat suggests, I'd vote for almost "starting again" with this one

**Comment by askvortsov1** [comment]
Alright, I'll close this one for now then.

**Pull request closed without merging**